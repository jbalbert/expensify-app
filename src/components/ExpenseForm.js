import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


export default class ExpenseForm extends React.Component {

    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note :  props.expense ? props.expense.note : '',
            amount : props.expense ? (props.expense.amount/100).toString() : '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error : '' 
        };
    }

  
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( () => ( {description}) );
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    }

    onAmountChange = (e) => {
        
        const amount = e.target.value;
        
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(() => ({amount}));
        }

    }

    onDateChange = (createdAt) => {

        if(createdAt){
            this.setState(() => ({createdAt}));
        }
       
    }

    onFocusedChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({error : 'Please Provide Description or Amount.'}));
           
        } else {
            this.setState(() => ({error : ''}));
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount, 10) * 100,
                createdAt : moment(this.state.createdAt).valueOf(),
                note : this.state.note
             });
        }

    }
    render() {
        return (
            <div>

               { this.state.error && <p> {this.state.error} </p> }
               <form onSubmit={this.onSubmit} >
                <input 
                type='text' 
                placeholder='Description' 
                value={this.state.description} 
                onChange={this.onDescriptionChange}
                autoFocus /> <br/> <br/>
                <input 
                type='text' 
                placeholder='Amount'
                value ={this.state.amount}
                onChange={this.onAmountChange}
                /> <br/> <br/>
                <SingleDatePicker
                    date = {this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.calendarFocused}
                    onFocusChange = {this.onFocusedChange}
                    numberOfMonths = {1}
                    isOutsideRange = {(day) => false}
                />
                <br/> <br/>
                <textarea placeholder='Add a note for your expense (optional)'
                    value={this.state.note}
                    onChange={this.onNoteChange}
                > 
                </textarea> <br/> <br/>
                <button> Add Expense </button>
               </form>
            </div>
        );  

    }

}