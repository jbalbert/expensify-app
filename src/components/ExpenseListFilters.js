import React from 'react';
import {connect} from 'react-redux';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filter';
import {DateRangePicker} from 'react-dates';
import uuid from 'uuid';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused : null
    };

    onDatesChange = ({startDate, endDate}) =>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    
    onSortChange = (e) => {
        const type = e.target.value;
        if(type === 'date'){
             this.props.sortByDate();
        } else if (type === 'amount'){
             this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <label>Filter Text : </label>
                <input 
                    type="text" defaultValue={this.props.filters.text} onChange={this.onTextChange}/>
                <label> Sort By : </label>
                <select 
                value={this.props.filters.sortBy} 
                onChange={this.onSortChange}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} 
                    endDate={this.props.filters.endDate} 
                    onDatesChange={this.onDatesChange} 
                    focusedInput={this.state.calendarFocused} 
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    startDateId = {uuid()}
                    endDateId = {uuid()}
                />
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filters : state.filters
    }
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : (text) => dispatch(setTextFilter(text)),
    sortByDate : () => dispatch(sortByDate()),
    sortByAmount : () => dispatch(sortByAmount),
    setStartDate : (startDate) => dispatch(setStartDate(startDate)),
    setEndDate : (endDate) => dispatch(setEndDate(endDate))

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);