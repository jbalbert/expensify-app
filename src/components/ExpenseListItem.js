import React from 'react';
//import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, createdAt, note}) => (
    <div>
        <br/> <br/> 
        <div> Expense ID : <Link to={`/edit/${id}`}> {id} </Link> </div> <br/>
        <div> Description : {description}  </div>  <br/>
        <div> Amount : {amount}  </div>  <br/>
        <div> Note : {note}  </div>  <br/>
        <div> CreatedAt : {createdAt}  </div> <br/> 
    </div>
);


//export default connect()(ExpenseListItem);

export default ExpenseListItem ;