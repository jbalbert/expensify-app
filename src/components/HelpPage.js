import React from 'react';

const HelpPage = () => (
    <div> This is a new help route</div>
);

export default HelpPage;