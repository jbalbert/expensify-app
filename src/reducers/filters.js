import moment from 'moment';

const filtersReducerDefaultState = {
    text : "",
    sortBy : 'date',
    startDate : moment().startOf('month'),
    endDate : moment().endOf('month')
};


const filtersReducer = (state = filtersReducerDefaultState,action) => {
    switch(action.type){

        case 'SET_TEXT_FILTER' : 

            return {
                ...state,
                text : action.text
            }
        case 'SET_AMOUNT_FILTER':
            return {
                ...state,
                sortBy : 'amount'
            }
        case 'SET_DATE_FILTER':
            return {
                ...state,
                sortBy : 'date'
            }
        case 'SET_START_DATE_FILTER':
            return {
                ...state,
                startDate : action.startDate
            }
        case 'SET_END_DATE_FILTER' : 
            return {
                ...state,
                endDate : action.endDate
            }
        
        default:
            return state;
    }
};


export default filtersReducer;