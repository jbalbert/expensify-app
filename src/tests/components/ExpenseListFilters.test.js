import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseListFilters} from '../../components/ExpenseListFilters';
import {filters, altfilters} from '../fixtures/filters';
import moment  from 'moment';
import 'react-dates/initialize';

let setTextFilter, sortByDate, sortByAmount, setStartDate,setEndDate, history, wrapper;

beforeEach(() => {
    setTextFilter = jest.fn();
    sortByDate = jest.fn();
    sortByAmount = jest.fn();
    setStartDate = jest.fn();
    setEndDate = jest.fn();
    history = {push : jest.fn()};
    wrapper = shallow(<ExpenseListFilters 
        filters={filters}
        setTextFilter={setTextFilter}
        sortByDate={sortByDate} 
        sortByAmount={sortByAmount} 
        setStartDate={setStartDate} 
        setEndDate={setEndDate} 
        history={history}/>);
});

test('should render expense list filters', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list filters with alt data', () => {
    wrapper.setProps({
       filters: altfilters
    })
    expect(wrapper).toMatchSnapshot();
});

test('should handle text change', () => {

    const value = 'moving';
    wrapper.find('input').simulate('change', {
        target :{
            value
        }
    });
    expect(setTextFilter).toHaveBeenLastCalledWith(value);
});

test('should handle sort by date', () => {

    wrapper.setProps({
        filters: altfilters
    });

    const value = 'date';

    wrapper.find('select').simulate('change', {
        target :{
            value
        }
    });
    expect(sortByDate).toHaveBeenLastCalledWith();

});

test('should handle sort by amount', () => {

    const value = 'amount';

    wrapper.find('select').simulate('change', {
        target :{
            value
        }
    });
    expect(sortByAmount).toHaveBeenLastCalledWith();
});


test('should handle date change', () => {

    const startDate = moment(0).add(4, 'years');
    const endDate = moment(0).add(8, 'years');

    wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')({
        startDate,
        endDate
    });

    expect(setStartDate).toHaveBeenLastCalledWith(startDate);
    expect(setEndDate).toHaveBeenLastCalledWith(endDate);
});


test('should handle focus change', () => {

    const calendarFocused = 'endDate';

    wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')(calendarFocused
    );

    expect(wrapper.state('calendarFocused')).toBe( calendarFocused);
});


