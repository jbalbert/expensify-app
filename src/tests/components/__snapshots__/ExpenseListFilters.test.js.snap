// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list filters 1`] = `
<div>
  <label>
    Filter Text : 
  </label>
  <input
    defaultValue=""
    onChange={[Function]}
    type="text"
  />
  <label>
     Sort By : 
  </label>
  <select
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={null}
    endDateId="a27b64e8-61c7-404a-a326-057c37acb1d5"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={null}
    startDateId="99ce9a4b-5560-40ba-94ab-0de24c97b3e8"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;

exports[`should render expense list filters with alt data 1`] = `
<div>
  <label>
    Filter Text : 
  </label>
  <input
    defaultValue="bills"
    onChange={[Function]}
    type="text"
  />
  <label>
     Sort By : 
  </label>
  <select
    onChange={[Function]}
    value="amount"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker)
    anchorDirection="left"
    appendToBody={false}
    block={false}
    calendarInfoPosition="bottom"
    customArrowIcon={null}
    customCloseIcon={null}
    customInputIcon={null}
    dayPickerNavigationInlineStyles={null}
    daySize={39}
    disableScroll={false}
    disabled={false}
    displayFormat={[Function]}
    enableOutsideDays={false}
    endDate={"1970-01-04T00:00:00.000Z"}
    endDateId="f433d13e-8314-476b-8b32-4a54b5d58520"
    endDatePlaceholderText="End Date"
    firstDayOfWeek={null}
    focusedInput={null}
    hideKeyboardShortcutsPanel={false}
    horizontalMargin={0}
    initialVisibleMonth={null}
    inputIconPosition="before"
    isDayBlocked={[Function]}
    isDayHighlighted={[Function]}
    isOutsideRange={[Function]}
    isRTL={false}
    keepFocusOnInput={false}
    keepOpenOnDateSelect={false}
    minimumNights={1}
    monthFormat="MMMM YYYY"
    navNext={null}
    navPosition="navPositionTop"
    navPrev={null}
    noBorder={false}
    numberOfMonths={1}
    onClose={[Function]}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    onNextMonthClick={[Function]}
    onPrevMonthClick={[Function]}
    openDirection="down"
    orientation="horizontal"
    phrases={
      Object {
        "calendarLabel": "Calendar",
        "chooseAvailableEndDate": [Function],
        "chooseAvailableStartDate": [Function],
        "clearDates": "Clear Dates",
        "closeDatePicker": "Close",
        "dateIsSelected": [Function],
        "dateIsSelectedAsEndDate": [Function],
        "dateIsSelectedAsStartDate": [Function],
        "dateIsUnavailable": [Function],
        "enterKey": "Enter key",
        "escape": "Escape key",
        "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
        "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
        "homeEnd": "Home and end keys",
        "jumpToNextMonth": "Move forward to switch to the next month.",
        "jumpToPrevMonth": "Move backward to switch to the previous month.",
        "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
        "keyboardShortcuts": "Keyboard Shortcuts",
        "leftArrowRightArrow": "Right and left arrow keys",
        "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
        "moveFocusByOneMonth": "Switch months.",
        "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
        "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
        "openThisPanel": "Open this panel.",
        "pageUpPageDown": "page up and page down keys",
        "questionMark": "Question mark",
        "returnFocusToInput": "Return to the date input field.",
        "roleDescription": "datepicker",
        "selectFocusedDate": "Select the date in focus.",
        "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
        "upArrowDownArrow": "up and down arrow keys",
      }
    }
    readOnly={false}
    regular={false}
    renderCalendarInfo={null}
    renderDayContents={null}
    renderMonthElement={null}
    renderMonthText={null}
    renderWeekHeaderElement={null}
    reopenPickerOnClearDates={false}
    required={false}
    screenReaderInputMessage=""
    showClearDates={true}
    showDefaultInputIcon={false}
    small={false}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="45b20ebe-6b5d-468f-9222-5ff94735af31"
    startDatePlaceholderText="Start Date"
    verticalHeight={null}
    verticalSpacing={22}
    weekDayFormat="dd"
    withFullScreenPortal={false}
    withPortal={false}
  />
</div>
`;
